### PR TITLE
feat(ui): add hostname validation for all node forms

### PR DESCRIPTION
**Pull request by huwshimi**
## Done

- Update all node forms to use the same hostname validation.

### MAAS deployment

To run this branch you will need access to one of the following MAAS deployments:

- [Bolla](/HACKING.md#bolla)
- [A development MAAS](/HACKING.md#development-deployment)
- [A local snap MAAS](/HACKING.md#snap-deployment) (this will not usually have machines)

### Running the branch

You can run this branch by:

- Serving with [dotrun](/HACKING.md#maas-ui-development-setup)
- [Building in a development MAAS](/HACKING.md#running-maas-ui-from-a-development-maas)

### QA steps

- Check that the add machine, add device etc. forms validate the hostname length and characters.

## Fixes

Fixes: #3806.

### CODE REVIEW

File: ui/src/app/base/validation.test.ts [hunk 1]
```diff
@@ -0,0 +1,21 @@
+import { hostnamePattern } from "./validation";
+
+describe("hostname validation", () => {
```
Comment (anthonydillon) [inline ui/src/app/base/validation.test.ts:3]: Could you add length tests for coverage and check the error label is returned as expected?

File: ui/src/app/base/validation.ts [hunk 2]
```diff
@@ -25,3 +27,19 @@ export const RANGE_REGEX = /^\d{1,3}(-\d{1,3})?(,\s*(\d{1,3}(-\d{1,3})?))*$/;
...
 export const TAG_NAME_REGEX = /^[a-zA-Z0-9_-]+$/;
+
+/**
+ * Validate host name string (only alphanumeric or dash and does not start or
+ * end with a dash)
+ */
+export const hostnamePattern =
+  /^([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9])*$/;
+
+export enum HostnameValidationLabel {
```
Comment (anthonydillon) [inline ui/src/app/base/validation.ts:38]: Might be nice to add an additional check for strings starting with a dash and return a label explaining that could be the cause of the error?

**Review by anthonydillon — APPROVED** [review]
LGTM with a few suggestions in the code

File: ui/src/app/base/validation.test.ts (hunk 1 above)
Comment (huwshimi) [reply ui/src/app/base/validation.test.ts]: Done.

File: ui/src/app/base/validation.ts (hunk 2 above)
Comment (huwshimi) [reply ui/src/app/base/validation.ts:31]: Done.

**Comment by huwshimi** [comment]
@anthonydillon thanks for the comments, I've made quite a few changes the validation and tests so could probably use another quick code review.

**Review by anthonydillon — APPROVED** [review]
Much nicer @huwshimi thanks for continuing to improve this PR

**Pull request merged**